vault-tool/src/Network/VaultTool/Types.hs

```diff
@@ -1,12 +1,36 @@
+{-# LANGUAGE GADTs #-}
 {-# LANGUAGE OverloadedStrings #-}
 
-module Network.VaultTool.Types where
+module Network.VaultTool.Types (
+    VaultAddress (..),
+    VaultAppRoleId (..),
+    VaultAppRoleSecretId (..),
+    VaultAppRoleSecretIdAccessor (..),
+    VaultAuthToken (..),
+    VaultConnection (..),
+    Authenticated,
+    Unauthenticated,
+    VaultException (..),
+    VaultMountedPath (..),
+    VaultSearchPath (..),
+    VaultSecretPath (..),
+    VaultUnsealKey (..),
+) where
 
 import Control.Exception (Exception)
 import Data.Aeson
 import Data.ByteString (ByteString)
 import Data.Text (Text)
 import qualified Data.ByteString.Lazy as BL
+import Network.HTTP.Client (Manager)
+
+data VaultConnection a where
+    UnauthenticatedVaultConnection :: Manager -> VaultAddress -> VaultConnection Unauthenticated
+    AuthenticatedVaultConnection :: Manager -> VaultAddress -> VaultAuthToken -> VaultConnection Authenticated
+
+data Unauthenticated
+
+data Authenticated
```

Comment:
We should also export this type, and add a Haddock.  My thinking is that if a type appears in the type signature of a package export, then it needs to be exported as well.

Reply:
2bf99a3268388fec77ead45d61e5b5f4251b6d96